KEYCLOAK-7907 Migration of model tests from old test suite.

mtrue01 opened the pull request:
This is a progress PR,  committting what I have done so far.

Commit:
KEYCLOAK-7907 Model tests migrated to new test suite.

Fixed versions

pdrozd commented:
@mtrue01 Can you update the commit message to include the issue number and the message?

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java:
```diff
@@ -0,0 +1,196 @@
+
+
+
+                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
```
Maybe this can be formatted?

mposolda replied:
And some of the empty lines can be removed?

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
```diff
@@ -0,0 +1,630 @@
+        return offlineSessions;
+    }
+
+    private static void resetSession(KeycloakSession session) {
```
This method can be probably entirely removed?

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
```diff
@@ -0,0 +1,630 @@
+        assertEquals(user.getUsername(), session.getLoginUsername());
+        assertEquals("form", session.getAuthMethod());
+        assertTrue(session.isRememberMe());
+        //assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
```
These checks should be uncommented

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
```diff
@@ -0,0 +1,630 @@
+            reloadState(sessionCrud);
+
+            UserSessionModel[] origSessions = createSessions(currentSession);
+            sessionCrud.getTransactionManager().commit();
```
As mentioned in the email, these calls to "sessionCrud.getTransactionManager().commit();" should be avoided. And nested transactions should be avoided too.

It's highly possible that even if nested transactions are passing with H2, they will probably fail with MySQL (and some other DBs).

In shortcut, if the old code contained something like this:
```
someMethodCalls1();

resetSession();

someMethodCalls2();
````

You can replace in the new with something like:

````
KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
                someMethodCalls1();
);

KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
                someMethodCalls2();
);
````
However not nested transactions.

mposolda replied:
This applies for many places in this class

mposolda reviewed (commented):
@mtrue01 Thanks for the PR. Added some inline comments.

mtrue01 commented:
I'll take a look and see what I can figure out.  As I said elsewhere, the issue I was running into was the scope of variables.  I will put together an example of what I am running into if I cannot figure it out.


> On Nov 14, 2018, at 2:35 PM, Marek Posolda <notifications@github.com> wrote:
> 
> @mposolda commented on this pull request.
> 
> @mtrue01 Thanks for the PR. Added some inline comments.
> 
> In testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java:
> 
> > +
> +        app1.addRole("app1Role1");
> +        app1.addScopeMapping(realm.getRole("role1"));
> +
> +        realm.addClient("cl1");
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testUsers(KeycloakSession session) {
> +
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
> +
> +
> +
> +                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
> Maybe this can be formatted?
> 
> In testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java:
> 
> > +
> +        app1.addRole("app1Role1");
> +        app1.addScopeMapping(realm.getRole("role1"));
> +
> +        realm.addClient("cl1");
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testUsers(KeycloakSession session) {
> +
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
> +
> +
> +
> +                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
> And some of the empty lines can be removed?
> 
> In testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
> 
> > +                        "org.keycloak.testsuite",
> +                        "org.keycloak.testsuite.model");
> +    }
> +
> +    private static Set<String> createOfflineSessionIncludeClientSessions(KeycloakSession session, UserSessionModel userSession) {
> +        Set<String> offlineSessions = new HashSet<>();
> +        UserSessionManager localManager = new UserSessionManager(session);
> +        for (AuthenticatedClientSessionModel clientSession : userSession.getAuthenticatedClientSessions().values()) {
> +            localManager.createOrUpdateOfflineSession(clientSession, userSession);
> +            offlineSessions.add(clientSession.getClient().getId());
> +        }
> +
> +        return offlineSessions;
> +    }
> +
> +    private static void resetSession(KeycloakSession session) {
> This method can be probably entirely removed?
> 
> In testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
> 
> > +        /*KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionReset) -> {
> +            currentSession.close();
> +            currentSession = sessionReset;
> +            realm = currentSession.realms().getRealm("test");
> +            sessionManager = new UserSessionManager(currentSession);
> +            persister = currentSession.getProvider(UserSessionPersisterProvider.class);
> +        });*/
> +    }
> +
> +    public static void assertSession(UserSessionModel session, UserModel user, String ipAddress, int started, int lastRefresh, String... clients) {
> +        assertEquals(user.getId(), session.getUser().getId());
> +        assertEquals(ipAddress, session.getIpAddress());
> +        assertEquals(user.getUsername(), session.getLoginUsername());
> +        assertEquals("form", session.getAuthMethod());
> +        assertTrue(session.isRememberMe());
> +        //assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
> These checks should be uncommented
> 
> In testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
> 
> > +                um.removeUser(realm, user2);
> +            }
> +        });
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testOfflineSessionsCrud(KeycloakSession session) {
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
> +            // Create some online sessions in infinispan
> +            int started = Time.currentTime();
> +
> +            reloadState(sessionCrud);
> +
> +            UserSessionModel[] origSessions = createSessions(currentSession);
> +            sessionCrud.getTransactionManager().commit();
> As mentioned in the email, these calls to "sessionCrud.getTransactionManager().commit();" should be avoided. And nested transactions should be avoided too.
> 
> It's highly possible that even if nested transactions are passing with H2, they will probably fail with MySQL (and some other DBs).
> 
> In shortcut, if the old code contained something like this:
> 
> someMethodCalls1();
> 
> resetSession();
> 
> someMethodCalls2();
> You can replace in the new with something like:
> 
> KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
>                 someMethodCalls1();
> );
> 
> KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
>                 someMethodCalls2();
> );
> However not nested transactions.
> 
> In testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
> 
> > +                um.removeUser(realm, user2);
> +            }
> +        });
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testOfflineSessionsCrud(KeycloakSession session) {
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
> +            // Create some online sessions in infinispan
> +            int started = Time.currentTime();
> +
> +            reloadState(sessionCrud);
> +
> +            UserSessionModel[] origSessions = createSessions(currentSession);
> +            sessionCrud.getTransactionManager().commit();
> This applies for many places in this class
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

mtrue01 commented:
I have annotated UserSessionProviderOfflineTest.java with my reasoning
behind why I used the nested transactions,   perhaps someone can point me
at a better way.

The thrust of it is that an object does not update properly without commit,
and the commit messes with the transaction manager.   Ending the block as
Marek P. had suggested causes the variable offLineSessions to go out of
scope.  offLineSessions is a hashmap and it is populated and then updated a
few times between line 201 and 220,  so having it go out of scope means
that any changes you've made to it are lost.  My way was to try and keep
the variable in scope and still be able to use the transaction system Marek
P. had put together because it got me around a transaction error.

If there is no commit between the creation of offLineSessions at 201 and
the for loop at 220,  then offlineSessions,size()

All the comments I have added annotating this all begin and end with *****
The first is on line 201:
https://github.com/mtrue01/keycloak/blob/633cd68a0be6b94226367e57ff751e66807cb60c/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java#L201

and the annotations continue over the next 20 lines or so to the conclusion
at line 220:
https://github.com/mtrue01/keycloak/blob/633cd68a0be6b94226367e57ff751e66807cb60c/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java#L220

Any help with this would be greatly appreciated.

--Mark T.

On Wed, Nov 14, 2018 at 2:35 PM Marek Posolda <notifications@github.com>
wrote:

> *@mposolda* commented on this pull request.
>
> @mtrue01 <https://github.com/mtrue01> Thanks for the PR. Added some
> inline comments.
> ------------------------------
>
> In
> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java
> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233587978>:
>
> > +
> +        app1.addRole("app1Role1");
> +        app1.addScopeMapping(realm.getRole("role1"));
> +
> +        realm.addClient("cl1");
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testUsers(KeycloakSession session) {
> +
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
> +
> +
> +
> +                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
>
> Maybe this can be formatted?
> ------------------------------
>
> In
> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java
> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233588136>:
>
> > +
> +        app1.addRole("app1Role1");
> +        app1.addScopeMapping(realm.getRole("role1"));
> +
> +        realm.addClient("cl1");
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testUsers(KeycloakSession session) {
> +
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
> +
> +
> +
> +                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
>
> And some of the empty lines can be removed?
> ------------------------------
>
> In
> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233588284>:
>
> > +                        "org.keycloak.testsuite",
> +                        "org.keycloak.testsuite.model");
> +    }
> +
> +    private static Set<String> createOfflineSessionIncludeClientSessions(KeycloakSession session, UserSessionModel userSession) {
> +        Set<String> offlineSessions = new HashSet<>();
> +        UserSessionManager localManager = new UserSessionManager(session);
> +        for (AuthenticatedClientSessionModel clientSession : userSession.getAuthenticatedClientSessions().values()) {
> +            localManager.createOrUpdateOfflineSession(clientSession, userSession);
> +            offlineSessions.add(clientSession.getClient().getId());
> +        }
> +
> +        return offlineSessions;
> +    }
> +
> +    private static void resetSession(KeycloakSession session) {
>
> This method can be probably entirely removed?
> ------------------------------
>
> In
> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233588501>:
>
> > +        /*KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionReset) -> {
> +            currentSession.close();
> +            currentSession = sessionReset;
> +            realm = currentSession.realms().getRealm("test");
> +            sessionManager = new UserSessionManager(currentSession);
> +            persister = currentSession.getProvider(UserSessionPersisterProvider.class);
> +        });*/
> +    }
> +
> +    public static void assertSession(UserSessionModel session, UserModel user, String ipAddress, int started, int lastRefresh, String... clients) {
> +        assertEquals(user.getId(), session.getUser().getId());
> +        assertEquals(ipAddress, session.getIpAddress());
> +        assertEquals(user.getUsername(), session.getLoginUsername());
> +        assertEquals("form", session.getAuthMethod());
> +        assertTrue(session.isRememberMe());
> +        //assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
>
> These checks should be uncommented
> ------------------------------
>
> In
> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233589663>:
>
> > +                um.removeUser(realm, user2);
> +            }
> +        });
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testOfflineSessionsCrud(KeycloakSession session) {
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
> +            // Create some online sessions in infinispan
> +            int started = Time.currentTime();
> +
> +            reloadState(sessionCrud);
> +
> +            UserSessionModel[] origSessions = createSessions(currentSession);
> +            sessionCrud.getTransactionManager().commit();
>
> As mentioned in the email, these calls to
> "sessionCrud.getTransactionManager().commit();" should be avoided. And
> nested transactions should be avoided too.
>
> It's highly possible that even if nested transactions are passing with H2,
> they will probably fail with MySQL (and some other DBs).
>
> In shortcut, if the old code contained something like this:
>
> someMethodCalls1();
>
> resetSession();
>
> someMethodCalls2();
>
> You can replace in the new with something like:
>
> KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
>                 someMethodCalls1();
> );
>
> KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
>                 someMethodCalls2();
> );
>
> However not nested transactions.
> ------------------------------
>
> In
> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233589844>:
>
> > +                um.removeUser(realm, user2);
> +            }
> +        });
> +    }
> +
> +    @Test
> +    @ModelTest
> +    public void testOfflineSessionsCrud(KeycloakSession session) {
> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
> +            // Create some online sessions in infinispan
> +            int started = Time.currentTime();
> +
> +            reloadState(sessionCrud);
> +
> +            UserSessionModel[] origSessions = createSessions(currentSession);
> +            sessionCrud.getTransactionManager().commit();
>
> This applies for many places in this class
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keycloak/keycloak/pull/5696#pullrequestreview-175038798>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZDqJFBEY9IVxs0gT9N3PMF_l9WXyeqmks5uvHCUgaJpZM4YQusO>
> .
>

mtrue01 commented:
On Wed, Nov 14, 2018 at 3:41 PM Mark True <mtrue@redhat.com> wrote:

> I have annotated UserSessionProviderOfflineTest.java with my reasoning
> behind why I used the nested transactions,   perhaps someone can point me
> at a better way.
>
> The thrust of it is that an object does not update properly without
> commit, and the commit messes with the transaction manager.   Ending the
> block as Marek P. had suggested causes the variable offLineSessions to go
> out of scope.  offLineSessions is a hashmap and it is populated and then
> updated a few times between line 201 and 220,  so having it go out of scope
> means that any changes you've made to it are lost.  My way was to try and
> keep the variable in scope and still be able to use the transaction system
> Marek P. had put together because it got me around a transaction error.
>
> If there is no commit between the creation of offLineSessions at 201 and
> the for loop at 220,  then offlineSessions,size()
>

This was meant to say:
If there is no commit between the creation of offLineSessions at 201 and
the for loop at 220,  then offlineSessions returns with a size of zero and
the for loop is skipped.


>
> All the comments I have added annotating this all begin and end with *****
> The first is on line 201:
>
> https://github.com/mtrue01/keycloak/blob/633cd68a0be6b94226367e57ff751e66807cb60c/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java#L201
>
> and the annotations continue over the next 20 lines or so to the
> conclusion at line 220:
>
> https://github.com/mtrue01/keycloak/blob/633cd68a0be6b94226367e57ff751e66807cb60c/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java#L220
>
> Any help with this would be greatly appreciated.
>
> --Mark T.
>
> On Wed, Nov 14, 2018 at 2:35 PM Marek Posolda <notifications@github.com>
> wrote:
>
>> *@mposolda* commented on this pull request.
>>
>> @mtrue01 <https://github.com/mtrue01> Thanks for the PR. Added some
>> inline comments.
>> ------------------------------
>>
>> In
>> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java
>> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233587978>:
>>
>> > +
>> +        app1.addRole("app1Role1");
>> +        app1.addScopeMapping(realm.getRole("role1"));
>> +
>> +        realm.addClient("cl1");
>> +    }
>> +
>> +    @Test
>> +    @ModelTest
>> +    public void testUsers(KeycloakSession session) {
>> +
>> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
>> +
>> +
>> +
>> +                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
>>
>> Maybe this can be formatted?
>> ------------------------------
>>
>> In
>> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/MultipleRealmsTest.java
>> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233588136>:
>>
>> > +
>> +        app1.addRole("app1Role1");
>> +        app1.addScopeMapping(realm.getRole("role1"));
>> +
>> +        realm.addClient("cl1");
>> +    }
>> +
>> +    @Test
>> +    @ModelTest
>> +    public void testUsers(KeycloakSession session) {
>> +
>> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionTU) -> {
>> +
>> +
>> +
>> +                RealmModel realm1 = sessionTU.realms().createRealm("id1", "realm1");
>>
>> And some of the empty lines can be removed?
>> ------------------------------
>>
>> In
>> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
>> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233588284>:
>>
>> > +                        "org.keycloak.testsuite",
>> +                        "org.keycloak.testsuite.model");
>> +    }
>> +
>> +    private static Set<String> createOfflineSessionIncludeClientSessions(KeycloakSession session, UserSessionModel userSession) {
>> +        Set<String> offlineSessions = new HashSet<>();
>> +        UserSessionManager localManager = new UserSessionManager(session);
>> +        for (AuthenticatedClientSessionModel clientSession : userSession.getAuthenticatedClientSessions().values()) {
>> +            localManager.createOrUpdateOfflineSession(clientSession, userSession);
>> +            offlineSessions.add(clientSession.getClient().getId());
>> +        }
>> +
>> +        return offlineSessions;
>> +    }
>> +
>> +    private static void resetSession(KeycloakSession session) {
>>
>> This method can be probably entirely removed?
>> ------------------------------
>>
>> In
>> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
>> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233588501>:
>>
>> > +        /*KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionReset) -> {
>> +            currentSession.close();
>> +            currentSession = sessionReset;
>> +            realm = currentSession.realms().getRealm("test");
>> +            sessionManager = new UserSessionManager(currentSession);
>> +            persister = currentSession.getProvider(UserSessionPersisterProvider.class);
>> +        });*/
>> +    }
>> +
>> +    public static void assertSession(UserSessionModel session, UserModel user, String ipAddress, int started, int lastRefresh, String... clients) {
>> +        assertEquals(user.getId(), session.getUser().getId());
>> +        assertEquals(ipAddress, session.getIpAddress());
>> +        assertEquals(user.getUsername(), session.getLoginUsername());
>> +        assertEquals("form", session.getAuthMethod());
>> +        assertTrue(session.isRememberMe());
>> +        //assertTrue(session.getStarted() >= started - 1 && session.getStarted() <= started + 1);
>>
>> These checks should be uncommented
>> ------------------------------
>>
>> In
>> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
>> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233589663>:
>>
>> > +                um.removeUser(realm, user2);
>> +            }
>> +        });
>> +    }
>> +
>> +    @Test
>> +    @ModelTest
>> +    public void testOfflineSessionsCrud(KeycloakSession session) {
>> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
>> +            // Create some online sessions in infinispan
>> +            int started = Time.currentTime();
>> +
>> +            reloadState(sessionCrud);
>> +
>> +            UserSessionModel[] origSessions = createSessions(currentSession);
>> +            sessionCrud.getTransactionManager().commit();
>>
>> As mentioned in the email, these calls to
>> "sessionCrud.getTransactionManager().commit();" should be avoided. And
>> nested transactions should be avoided too.
>>
>> It's highly possible that even if nested transactions are passing with
>> H2, they will probably fail with MySQL (and some other DBs).
>>
>> In shortcut, if the old code contained something like this:
>>
>> someMethodCalls1();
>>
>> resetSession();
>>
>> someMethodCalls2();
>>
>> You can replace in the new with something like:
>>
>> KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
>>                 someMethodCalls1();
>> );
>>
>> KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud2) -> {
>>                 someMethodCalls2();
>> );
>>
>> However not nested transactions.
>> ------------------------------
>>
>> In
>> testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java
>> <https://github.com/keycloak/keycloak/pull/5696#discussion_r233589844>:
>>
>> > +                um.removeUser(realm, user2);
>> +            }
>> +        });
>> +    }
>> +
>> +    @Test
>> +    @ModelTest
>> +    public void testOfflineSessionsCrud(KeycloakSession session) {
>> +        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), (KeycloakSession sessionCrud) -> {
>> +            // Create some online sessions in infinispan
>> +            int started = Time.currentTime();
>> +
>> +            reloadState(sessionCrud);
>> +
>> +            UserSessionModel[] origSessions = createSessions(currentSession);
>> +            sessionCrud.getTransactionManager().commit();
>>
>> This applies for many places in this class
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/keycloak/keycloak/pull/5696#pullrequestreview-175038798>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AZDqJFBEY9IVxs0gT9N3PMF_l9WXyeqmks5uvHCUgaJpZM4YQusO>
>> .
>>
>

Commit:
foo

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionProviderOfflineTest.java:
```diff
@@ -0,0 +1,584 @@
+
+    @Test
+    @ModelTest
+    public void testExpired(KeycloakSession session) {
```
The "testExpired" test was updated in the master due some offlineSessions preloading work. Could you try to rebase and doublecheck that new bits added to the test in the meantime are reflected in the migrated test too?

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/model/UserSessionPersisterProviderTest.java:
```diff
@@ -0,0 +1,525 @@
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class UserSessionPersisterProviderTest extends AbstractTestRealmKeycloakTest {
```
There were some changes in the test due the offline sessions. Some new methods added (testUpdateTimestamps) and some logic changed in few existing tests. Could you try to make sure that these new tests are migrated as well and it's not lost by migration?

mposolda reviewed (changes requested):
@mtrue01 Thanks for all the updates. Looks quite good to me. I left few comments. In summary:
- Some minor formatting issues like more spaces etc (hopefully should be easy for you to fix)
- Some methods changed in the meantime in the UserSessionPersisterProviderTest and UserSessionProviderOfflineTest. Sorry for the complications (it's me who added those), however will be good  to ensure that new tests are not lost during migration
- You commented some assert tests in one of the places (I think it's the check for "started" and "lastSessionRefresh"). Could you uncomment and see if tests passing with it uncommented?
- In one of the tests, you used 1 transaction instead of 3. Maybe it's fine, but will be good to re-test with MySQL. If it's failing, it is probably needed to add back 3 transactions.
- UserSessionPersisterProviderTest can be removed from the old testsuite? It seems you removed most of the tests from the old testsuite, but not this one? Or is it still needed (referenced from some other test?)

mposolda commented:
Ah, one final note. Once you address the comments, we will probably need to trigger DB matrix job. There is a chance that some tests will fail for some DB as it seems the old testsuite wasn't executed with DB matrix for the long time :( But maybe false alarm, we will see...

Commit:
Merge branch 'master' of https://github.com/keycloak/keycloak

mabartos commented:
This PR is replaced by [PR#5898](https://github.com/keycloak/keycloak/pull/5898)

mposolda commented:
@mabartos Thanks!
